packages/@glimmer/application/src/components/component-managers/template-only.ts

```diff
@@ -0,0 +1,173 @@
+  wrapped: false,
+};
+
+export class ComponentStateBucket {
+  public name: string;
+
+  constructor(public args: CapturedArguments) {
+    this.args = args;
+  }
+}
+
+const EMPTY_SELF = new RootReference(null);
+
+/**
+ * For performance reasons, we want to avoid instantiating component buckets for
+ * components that don't have an associated component class that we would need
+ * instantiate and invoke lifecycle hooks on.
+ *
+ * In development mode, however, we need to track some state about the component
+ * in order to produce more useful error messages. This
+ * TemplateOnlyComponentDebugBucket is only created in development mode to hold
+ * that state.
+ */
+export class TemplateOnlyComponentDebugBucket {
+  constructor(public definition: TemplateOnlyComponentDefinition) {}
+}
+
+export interface ExtendedTemplateMeta {
+  specifier: string;
+  managerId?: string;
+}
+
+export interface CompilableRuntimeResolver extends RuntimeResolver<ExtendedTemplateMeta> {
+  compileTemplate(name: string, layout: Option<number>): Invocation;
+}
+
+export interface EnvironmentWithOwner extends Environment {
+  getOwner(): Owner;
+  setOwner(obj: Object, owner: Owner): void;
+}
+
+export default class TemplateOnlyComponentManager
+  implements
+    VMComponentManager<TemplateOnlyComponentDebugBucket | null, ComponentDefinitionState>,
+    WithJitStaticLayout<
+      TemplateOnlyComponentDebugBucket | null,
+      TemplateOnlyComponentDefinitionState,
+      JitRuntimeResolver
+    >,
+    WithAotStaticLayout<
+      TemplateOnlyComponentDebugBucket | null,
+      TemplateOnlyComponentDefinitionState,
+      AotRuntimeResolver
+    > {
+  static create(): TemplateOnlyComponentManager {
+    return new TemplateOnlyComponentManager();
+  }
+
+  getCapabilities(): ComponentCapabilities {
+    return CAPABILITIES;
+  }
+
+  getJitStaticLayout({ definition }: TemplateOnlyComponentDefinitionState): CompilableProgram {
+    return definition.template.asLayout();
+  }
+
+  getAotStaticLayout({ definition }: TemplateOnlyComponentDefinitionState): Invocation {
+    const { handle, symbolTable } = definition;
+    return {
+      handle,
+      symbolTable
+    };
+  }
+
+  create(
+    _env: Environment,
+    state: TemplateOnlyComponentDefinitionState
+  ): TemplateOnlyComponentDebugBucket | void {
+    // In development mode, save off state needed for error messages. This will
+    // get stripped in production mode and no bucket will be instantiated.
+    return DEBUG ? new TemplateOnlyComponentDebugBucket(state.definition) : null;
+  }
+
+  getSelf(bucket: TemplateOnlyComponentDebugBucket): PathReference {
+    return DEBUG ? new TemplateOnlyComponentDebugReference(bucket.definition.state.name) : EMPTY_SELF;
+  }
+
+  getTag(): Tag {
+    return CONSTANT_TAG;
+  }
+
+  didCreateElement() {}
+  didRenderLayout() {}
+  didCreate() {}
+  update() {}
+  didUpdateLayout() {}
+  didUpdate() {}
+  getDestructor(): Destroyable {
+    return NOOP_DESTROYABLE;
+  }
```

Comment:
Should be able to just return `null` here?

Reply:
Good call